### PR TITLE
bootstrap: (half) merge cp_r and cp_filtered

### PR DESCRIPTION
**Pull request by klensy**
Half-way merges cp_r and cp_filtered functions, as there some different points:

`cp_r` and `cp_filtered` difference:
skip running if dry_run => `cp_r`: yes, `cp_filtered`: no
removes target dir if it exist (i.e. can't be used to copy files into dst dir from multiple sources, (or no?)) => `cp_r`: no,  `cp_filtered`: yes

Currently this two places didn't changed.

### CODE REVIEW

**Comment by klensy** [comment]
Looking at 2 uses of `cp_filtered`, target directory created somewhere before and should be empty, so `remove_dir_all` (as second difference between cp_r and cp_filtered) can be removed.

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -1570,11 +1561,9 @@ impl Build {
             // Only copy file or directory if the filter function returns true
             if filter(&relative) {
                 if t!(f.file_type()).is_dir() {
-                    let _ = fs::remove_dir_all(&dst);
                     self.create_dir(&dst);
                     self.recurse_(&path, &dst, &relative, filter);
                 } else {
-                    let _ = fs::remove_file(&dst);
```
Comment (Mark-Simulacrum) [inline src/bootstrap/lib.rs:1577]: Can you say more about removing these lines? What is replacing them / why is it safe to do that?

Comment (klensy) [reply src/bootstrap/lib.rs:1577]: I've described why this line can be removed in commit's title:
`dont call remove_file in cp_filtered, as self.copy will call copy_internal which will call remove_file itself`

Hm, looks like i mean `recurse_` function, not `cp_filtered`

Comment (klensy) [reply src/bootstrap/lib.rs:1577]: Comment about `remove_dir_all`: https://github.com/rust-lang/rust/pull/111734#issuecomment-1556148879

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -1538,18 +1538,7 @@ impl Build {
         if self.config.dry_run() {
             return;
         }
```
Comment (klensy) [inline src/bootstrap/lib.rs:1540]: So this `dry_run` is only difference.

**Comment by Mark-Simulacrum** [comment]
> Looking at 2 uses of cp_filtered, target directory created somewhere before and should be empty, so remove_dir_all (as second difference between cp_r and cp_filtered) can be removed.

I don't think this is valid reasoning. We don't want to write this code against *current* usage, we want it to work against future usages as well -- which may have a destination directory that does exist.

File: src/bootstrap/lib.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/lib.rs:1540]: But we may break things if we remove it? I'm not sure we want to do this clean up, seems like these functions are distinct if similar today.

**Comment by JohnCSimon** [comment]
@klensy 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**